torch/csrc/cuda/CUDAPluggableAllocator.cpp

```diff
@@ -0,0 +1,218 @@
+    if (capture_destroy_fn_.has_value()) {
+      capture_destroy_fn_.value()(device, mempool_id);
+    }
+  }
+
+  virtual void* raw_alloc(size_t nbytes) {
+     return nullptr;
+  }
+  virtual void* raw_alloc_with_stream(size_t nbytes, cudaStream_t stream) {
+     return nullptr;
+  }
+
+  virtual void raw_delete(void* ptr) {
+  }
+
+  virtual void recordHistory(
+      bool enabled,
+      c10::cuda::CUDACachingAllocator::CreateContextFn context_recorder,
+      size_t alloc_trace_max_entries,
+      bool alloc_trace_record_context) {
+  }
+
+  virtual void attachOutOfMemoryObserver(c10::cuda::CUDACachingAllocator::OutOfMemoryObserver observer) {
+  }
+
+  virtual bool needsPoolSpecificPeerAccess() {
+    return false;
+  }
+
+  virtual std::string name() {
+    return "pluggable";
+  }
+
+protected:
+  std::function<void*(size_t, int, cudaStream_t)> alloc_fn_;
+  std::function<void(void*)> free_fn_;
+  // TODO Populate all these functions
+  c10::optional<std::function<void(int)>> init_fn_ = c10::nullopt;
+  c10::optional<std::function<void()>> reset_fn_ = c10::nullopt;
+  c10::optional<std::function<void*(void*, size_t*)>> base_alloc_fn_ = c10::nullopt;
+  c10::optional<std::function<void(void* ptr, cudaStream_t stream)>> record_stream_fn_ = c10::nullopt;
+  // CUDAGraph management
+  // TODO(ecastill) mempool id is needed here?
+  c10::optional<std::function<void(int, c10::cuda::CaptureId_t, c10::cuda::MempoolId_t)>> capture_begin_fn_ = c10::nullopt;
+  c10::optional<std::function<void(int, c10::cuda::CaptureId_t)>> capture_about_to_end_fn_ = c10::nullopt;
+  c10::optional<std::function<void(int, c10::cuda::CaptureId_t)>> capture_ended_fn_ = c10::nullopt;
+  c10::optional<std::function<void(int, c10::cuda::MempoolId_t)>> capture_destroy_fn_ = c10::nullopt;
+  // TODO(ecastill) should we add support for profiling functions? maybe in another PR
+};
+
+std::shared_ptr<c10::cuda::CUDACachingAllocator::CUDAAllocator> getCurrentAllocator() {
+  // Empty deleter to avoid double free, this is to avoid having additional variables to hold custom allocator
+  // and rely on `c10::cuda::CUDACachingAllocator::get()` as the only allocator
+  return std::shared_ptr<c10::cuda::CUDACachingAllocator::CUDAAllocator>(
+    c10::cuda::CUDACachingAllocator::get(), [](c10::cuda::CUDACachingAllocator::CUDAAllocator*){});
+}
+
+// TODO: add more functions in the argument
+std::shared_ptr<c10::cuda::CUDACachingAllocator::CUDAAllocator> createCustomAllocator(
+  std::function<void*(size_t, int, cudaStream_t)> alloc_fn,
+  std::function<void(void*)> free_fn) {
+  std::shared_ptr<c10::cuda::CUDACachingAllocator::CUDAAllocator> allocator(new CudaCustomAllocator(alloc_fn, free_fn));
+  allocator->init(device_count);
+  return allocator;
+}
+
+void changeCurrentAllocator(std::shared_ptr<c10::cuda::CUDACachingAllocator::CUDAAllocator> allocator) {
+  c10::cuda::CUDACachingAllocator::allocator.store(allocator.get());
```

Comment:
If the allocator no longer has references in Python, then the object is going to be destructed but still have a dangling reference here. Maybe add a static variable that holds on the shared pointer here in addition to setting the allocator.

Also it is unsafe to change the allocator after some allocations have happened because a `raw_alloc`'d pointer might later get passed to a `raw_free` of a different allocator.  The use of the `raw_` APIs is rare but are used in the codebase. What to do depends more on the use cases you might have in mind for this API. Ideally, we'd try to get rid of the raw_deletes and instead have them use the DataPtr interface so that even if the allocator is different at the point of free, the DataPtr still directs to the correct allocator. 

A simpler choice is to try to figure out if we have allocated any memory yet, and just refuse to change the allocator if we have.

Reply:
I went with your approach and forbid allocator changes after initialization. However, I would like to remove the `raw_` apis so we can dynamically change allocators in different parts of the application (p.e. used managed memory in one place and then resort to regular cuda mallocs).